jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeouts.java

```diff
@@ -0,0 +1,147 @@
+
+    @Override
+    public void onTimeoutExpired()
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} timeouts check", this);
+
+        long now = System.nanoTime();
+        long earliest = Long.MAX_VALUE;
+        // Reset the earliest timeout so we can expire again.
+        // A concurrent call to schedule(long) may lose an
+        // earliest value, but the corresponding entity will
+        // be seen during the iteration below.
+        earliestTimeout.set(earliest);
+
+        // Scan the entities to abort expired entities
+        // and to find the entity that expires the earliest.
+        Iterator<T> iterator = iterator();
+        while (iterator.hasNext())
+        {
+            T expirable = iterator.next();
+            long expiresAt = expirable.getExpireNanoTime();
+            if (expiresAt == -1)
+                continue;
+            if (expiresAt <= now)
+            {
+                if (onExpired(expirable))
+                    iterator.remove();
+                continue;
+            }
+            earliest = Math.min(earliest, expiresAt);
+        }
+
+        if (earliest < Long.MAX_VALUE)
+            schedule(earliest);
+    }
+
+    /**
+     * <p>Manages the timeout of a new entity.</p>
+     *
+     * @param expirable the new entity to manage the timeout for
+     */
+    public void schedule(T expirable)
+    {
+        long expiresAt = expirable.getExpireNanoTime();
+        if (expiresAt < Long.MAX_VALUE)
+            schedule(expiresAt);
+    }
+
+    private void schedule(long expiresAt)
+    {
+        // Schedule a timeout for the earliest entity that may expire.
+        // When the timeout expires, scan the entities for the next
+        // earliest entity that may expire, and reschedule a new timeout.
+        long prevEarliest = earliestTimeout.getAndUpdate(t -> Math.min(t, expiresAt));
+        if (expiresAt < prevEarliest)
+        {
+            // A new entity expires earlier than previous entities, schedule it.
+            long delay = Math.max(0, expiresAt - System.nanoTime());
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} scheduling timeout in {} ms", this, TimeUnit.NANOSECONDS.toMillis(delay));
+            schedule(delay, TimeUnit.NANOSECONDS);
+        }
+    }
+
+    /**
+     * <p>An entity that may expire.</p>
+     */
+    public interface Expirable
+    {
+        /**
+         * <p>Returns the expiration time in nanoseconds.</p>
+         * <p>The value to return must be calculated taking into account {@link System#nanoTime()},
+         * for example:</p>
+         * {@code expireNanoTime = System.nanoTime() + timeoutNanos}
+         * <p>Returning {@link Long#MAX_VALUE} indicates that this entity does not expire.</p>
+         *
+         * @return the expiration time in nanoseconds, or {@link Long#MAX_VALUE} if this entity does not expire
+         */
+        public long getExpireNanoTime();
```

Comment:
```suggestion
        public long getExpireAtNanoTime();
```

Reply:
Uhm. 

Either my `expireNanoTime` where `expire` is noun-ish, or `expire*s*At` where `expire` is verb-ish (so needs the `s`).